File: frontend/src/types/balance.ts
export interface Balance {
    id: string;
    senderUserId: string;
    senderUsername: string;
    receiverUserId: string;
    receiverUsername: string;
    balance: string;
    currency: string;
}

export interface SettlementPreviewContribution {
    balanceId: string;
    sourceCurrency: string;
    sourceAmount: string;
    rate: string | null;
    previewAmount: string | null;
}

export interface SettlementPreview {
    currency: string;
    complete: boolean;
    netAmount: string | null;
    contributions: SettlementPreviewContribution[];
    missingRateCurrencies: string[];
}

export interface BalanceData {
    currency: string;
    currentUser: string;
    balances: Balance[];
    settlementPreview: SettlementPreview | null;
}

export enum SplitOption {
    Equally = "Equally",
    Unequally = "Unequally",
    YouHalf = "You-Half",
    YouFull = "You-Full",
    OtherHalf = "Other-Half",
    OtherFull = "Other-Full",
}

File: backend/types/balance.go
package types

import (
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

type Balance struct {
	ID             uuid.UUID
	SenderUserID   uuid.UUID
	ReceiverUserID uuid.UUID
	Share          decimal.Decimal
	GroupID        uuid.UUID
	CreateTime     time.Time
	IsOutdated     bool
	UpdateTime     time.Time
	IsSettled      bool
	SettledTime    time.Time
	Currency       string
}

type BalanceRsp struct {
	ID               uuid.UUID       `json:"id"`
	SenderUserID     uuid.UUID       `json:"senderUserId"`
	SenderUesrname   string          `json:"senderUsername"`
	ReceiverUserID   uuid.UUID       `json:"receiverUserId"`
	ReceiverUsername string          `json:"receiverUsername"`
	Balance          decimal.Decimal `json:"balance"`
	Currency         string          `json:"currency"`
}

type BalanceResponse struct {
	Currency          string             `json:"currency"`
	CurrentUser       string             `json:"currentUser"`
	Balances          []BalanceRsp       `json:"balances"`
	SettlementPreview *SettlementPreview `json:"settlementPreview"`
}

type SettlementPreviewContribution struct {
	BalanceID      uuid.UUID        `json:"balanceId"`
	SourceCurrency string           `json:"sourceCurrency"`
	SourceAmount   decimal.Decimal  `json:"sourceAmount"`
	Rate           *decimal.Decimal `json:"rate"`
	PreviewAmount  *decimal.Decimal `json:"previewAmount"`
}

type SettlementPreview struct {
	Currency              string                          `json:"currency"`
	Complete              bool                            `json:"complete"`
	NetAmount             *decimal.Decimal                `json:"netAmount"`
	Contributions         []SettlementPreviewContribution `json:"contributions"`
	MissingRateCurrencies []string                        `json:"missingRateCurrencies"`
}
